### PR TITLE
Add "How-To Create a Container" documentation

### PR DESCRIPTION
**Pull request by apoudret-microej**

### CODE REVIEW

File: MWT/how-to-container.rst [hunk 1]
```diff
@@ -1,6 +1,66 @@
...
+For example, the optimal width of a horizontal list is the sum of the optimal width of every child, and the optimal height is the maximum optimal height of all children.
+
+The container is responsible for computing the optimal size of every child. To do so, the ``computeChildOptimalSize()`` method should be called for every child.
+After this method is called, the optimal size of the child can be retrieved by calling ``getWidth()`` and ``getHeight()`` on the child widget.
+
+The ``Size`` parameter of the ``computeContentOptimalSize()`` method initially contains the size available for the container.
+An available width or height equal to ``Widget.NO_CONSTRAINT`` means that the optimal size should be computed without considering any restriction on the respective axis.
+Before the method returns, the size object should be set to the optimal size of the container.
+
+"blurry point": should the optimal size always be lower than the available size?
```
Comment (seonmej) [inline MWT/how-to-container.rst]: I want to say "no" ;-)
The optimal size of a label on one line is always the same whatever the available size (even if it is smaller).

File: MWT/how-to-container.rst [hunk 2]
```diff
@@ -1,6 +1,66 @@
...
+
+Laying out the children of the container
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``layOutChildren()`` method is called by the MWT framework in order to lay out every child of the container.
```
Comment (seonmej) [inline MWT/how-to-container.rst]: Perhaps explain the meaning of lay out here? `Meaning setting the bounds (x, y, width and height) of the children.`

File: MWT/how-to-container.rst [hunk 3]
```diff
@@ -1,6 +1,66 @@
...
+If a child is laid out outside the bounds of the container (partially or fully), only the part of the widget which is within the container bounds will be visible.
+
+The container is responsible for laying out each child. To do so, the ``layOutChild()`` method should be called for every child.
+Before this method is called, the optimal size of the child can be retrieved by calling ``getWidth()`` and ``getHeight()`` on the child widget.
+
+When laying out a child, its bounds have to be passed as parameter. The position will be interpreted as relative to the position of the container content.
```
Comment (seonmej) [inline MWT/how-to-container.rst]: I think that the `container content` needs to be explained: the boxes (outlines and border) are removed.

Comment (seonmej) [reply MWT/how-to-container.rst]: And the boxes must also be explained in the previous section (optimal size).

File: MWT/how-to-container.rst [hunk 4]
```diff
@@ -1,6 +1,66 @@
...
+This means that the ``(0, 0)`` coordinates represent the top-left pixel of the container content and the ``(contentWidth-1, contentHeight-1)`` coordinates represent the bottom-right pixel of the container content.
+
+Providing children manipulation APIs
+------------------------------------
+
+The ``Container`` class introduces protected APIs in order to manipulate the list of children of the container.
+These methods may be overriden in the container subclass in order to make these APIs available for the user.
```
Comment (seonmej) [inline MWT/how-to-container.rst]: Overriden and set as `public`?

File: MWT/how-to-container.rst [hunk 5]
```diff
@@ -1,6 +1,66 @@
...
+- ``changeChildIndex()``
+
+For example, the following snippet allows the user to call the ``addChild()`` method on the container:
+
+.. code-block:: Java
+
+	@Override
+	public void addChild(Widget child) {
+		super.addChild(child);
+	}
```
Comment (seonmej) [inline MWT/how-to-container.rst:93]: Another notion that can be added (maybe later) is the possibility to change the default children visibility by overriding `setShownChildren()` and using `setShownChild()`.

Comment (apoudret-microej) [reply MWT/how-to-container.rst:93]: i forgot this method, i will add a few lines

File: MWT/how-to-container.rst (hunk 1 above)
Comment (apoudret-microej) [reply MWT/how-to-container.rst]: hm currently no, StringPainter & ImagePainter call Math.min to get minimum value between text/image size and available size. i think we should decide and make it clear

File: MWT/how-to-container.rst (hunk 3 above)
Comment (apoudret-microej) [reply MWT/how-to-container.rst]: i think the notion of content should be explained in architecture page, in the boxing model section

File: MWT/how-to-container.rst (hunk 1 above)
Comment (seonmej) [reply MWT/how-to-container.rst]: Oh, indeed, I didn't saw that. In previous Label, the returned size was always the same and there was no issue. If the widget optimal size is bigger than the constraint, its will probably exceed its final bounds, that's all.

Comment (apoudret-microej) [reply MWT/how-to-container.rst]: yes that's right, i copy/pasted this behavior from ImagePainter because i thought it was nice
ok so this min() shouldn't be done anymore?

File: MWT/how-to-container.rst (hunk 3 above)
Comment (seonmej) [reply MWT/how-to-container.rst]: Of course, but I think that here, we should explain when the boxes are used.

File: MWT/how-to-container.rst (hunk 1 above)
Comment (seonmej) [reply MWT/how-to-container.rst]: I think it is useless, yes.

File: MWT/how-to-container.rst (hunk 5 above)
Comment (apoudret-microej) [reply MWT/how-to-container.rst:93]: done

File: MWT/how-to-container.rst (hunk 4 above)
Comment (apoudret-microej) [reply MWT/how-to-container.rst]: fixed

File: MWT/how-to-container.rst (hunk 1 above)
Comment (apoudret-microej) [reply MWT/how-to-container.rst]: fixed

File: MWT/how-to-container.rst (hunk 2 above)
Comment (apoudret-microej) [reply MWT/how-to-container.rst]: done

File: MWT/how-to-container.rst (hunk 3 above)
Comment (apoudret-microej) [reply MWT/how-to-container.rst]: done

File: MWT/how-to-event-dispatcher.rst [hunk 6]
```diff
@@ -6,6 +6,59 @@ Two use cases:
...
+
+Likewise, the ``dispose()`` method may be overridden to remove the event generator from the system pool of MicroUI:
+
+.. code-block:: Java
+
+	@Override
+	public void dispose() {
+		this.eventGenerator.removeFromSystemPool();
+	}
+
```
Comment (seonmej) [inline MWT/how-to-event-dispatcher.rst:62]: Maybe explain how to send custom events?

File: MWT/how-to-container.rst [hunk 7]
```diff
@@ -14,8 +14,7 @@ Computing the optimal size of the container
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 The ``computeContentOptimalSize()`` method is called by the MWT framework in order to know the optimal size of the container.
-The optimal size of the container should be big enough so that each child can be allocated a size at least as big as its own optimal size.
-For example, the optimal width of a horizontal list is the sum of the optimal width of every child, and the optimal height is the maximum optimal height of all children.
```
Comment (seonmej) [inline MWT/how-to-container.rst]: Why did you remove this example?

File: MWT/how-to-widget.rst [hunk 8]
```diff
@@ -1,6 +1,99 @@
...
+		Style style = getStyle();
+		g.setColor(style.getColor());
+		Painter.drawString(g, style.getFont(), this.text, 0, 0);
+	}
+
+Handling events
+---------------
+
+When a widget is created, it is disabled and it will not receive any event.
+A widget may be enabled or disabled by calling ``setEnabled()``. A common practice is to enable the widget is its constructor.
```
Comment (seonmej) [inline MWT/how-to-widget.rst]: `in its constructor`

File: MWT/how-to-widget.rst [hunk 9]
```diff
@@ -1,6 +1,99 @@
...
+ 
+Enabled widgets can handle events by overriding ``handleEvent()``. MicroUI event APIs may be used in order to know more information on the event, such as its type.
+ 
+For example, the following snippet prints a message when the widget receives an event:
```
Comment (seonmej) [inline MWT/how-to-widget.rst:61]: Also explain that the result of the method is whether or not the event is consumed.

File: MWT/how-to-widget.rst [hunk 10]
```diff
@@ -1,6 +1,99 @@
...
+	}
+
+Likewise, the ``onDetached()`` method may be overridden to close the image:
+
+.. code-block:: Java
+
+	@Override
+	protected void onDetached() {
+		this.image.close();
+	}
```
Comment (seonmej) [inline MWT/how-to-widget.rst:98]: An example for `onShown()` that starts an Animation? (or TimerTask or Thread.)

**Review by seonmej — CHANGES_REQUESTED** [review]

File: MWT/how-to-container.rst (hunk 7 above)
Comment (apoudret-microej) [reply MWT/how-to-container.rst]: instead i added a code snippet to compute the optimal size of a container. i chose a simple wrapper rather than a horizontal list to have a shorter example

File: MWT/how-to-background.rst [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+	@Override
+	public boolean isTransparent() {
+		return false;
+	}
+
+Applying the background on a graphics context
+---------------------------------------------
+
+The ``apply(GraphicsContext g, Size size)`` method is called by the MWT framework in order to render the background and to set the background color of subsequent drawings.
```
Comment (seonmej) [inline MWT/how-to-background.rst]: Indicate that setting the background color is not mandatory.

**Review by seonmej — CHANGES_REQUESTED** [review]

File: MWT/how-to-background.rst (hunk 11 above)
Comment (apoudret-microej) [reply MWT/how-to-background.rst]: then "set or remove"? i think doing either of them is mandatory

File: MWT/how-to-widget.rst (hunk 10 above)
Comment (apoudret-microej) [reply MWT/how-to-widget.rst:98]: done

File: MWT/how-to-widget.rst (hunk 8 above)
Comment (apoudret-microej) [reply MWT/how-to-widget.rst]: fixed

File: MWT/how-to-container.rst (hunk 7 above)
Comment (seonmej) [reply MWT/how-to-container.rst]: ok

File: MWT/how-to-background.rst (hunk 11 above)
Comment (seonmej) [reply MWT/how-to-background.rst]: Yes, good point!

File: MWT/how-to-event-dispatcher.rst (hunk 6 above)
Comment (apoudret-microej) [reply MWT/how-to-event-dispatcher.rst:62]: done (added to in subsection above)

File: MWT/how-to-background.rst (hunk 11 above)
Comment (apoudret-microej) [reply MWT/how-to-background.rst]: done
